service/isiService.go

```diff
@@ -784,44 +784,75 @@ func (svc *isiService) GetExportWithPathAndZone(ctx context.Context, path, acces
+	return svc.client.GetSnapshotIsiPath(ctx, isiPath, sourceSnapshotID, accessZone)
 }
 
-func (svc *isiService) isROVolumeFromSnapshot(isiPath string) bool {
-	if strings.Index(isiPath, constants.VolumeSnapshotsPath) == 0 {
-		return true
+func (svc *isiService) GetZoneByName(ctx context.Context, accessZone string) (*apiv1.IsiZone, error) {
+	zone, err := svc.client.GetZoneByName(ctx, accessZone)
+	return zone, err
+}
+
+func (svc *isiService) isROVolumeFromSnapshot(exportPath, accessZone string) bool {
+	isROVolFromSnapshot := false
+	if accessZone == "System" {
+		if strings.Index(exportPath, "/ifs/.snapshot") == 0 {
+			isROVolFromSnapshot = true
+		}
+	} else {
+		if strings.Index(exportPath, "/.snapshot") != -1 {
+			isROVolFromSnapshot = true
+		}
 	}
-	return false
+	return isROVolFromSnapshot
 }
 
-func (svc *isiService) GetSnapshotNameFromIsiPath(ctx context.Context, snapshotIsiPath string) (string, error) {
+func (svc *isiService) GetSnapshotNameFromIsiPath(ctx context.Context, snapshotIsiPath, accessZone, zonePath string) (string, error) {
 	// Fetch log handler
 	log := utils.GetRunIDLogger(ctx)
-
-	if !svc.isROVolumeFromSnapshot(snapshotIsiPath) {
+	var snapShotName string
+	if !svc.isROVolumeFromSnapshot(snapshotIsiPath, accessZone) {
 		log.Debugf("invalid snapshot isilon path- '%s'", snapshotIsiPath)
 		return "", fmt.Errorf("invalid snapshot isilon path")
 	}
-
 	// Snapshot isi path format /<ifs>/.snapshot/<snapshot_name>/<volume_path_without_ifs_prefix>
-	dirs := strings.Split(snapshotIsiPath, "/")
+	//Non System Access Zone /<ifs>/<csi_zone_base_path>/.snapshot/<snapshot_name>/<volume_path_without_ifs_prefix>
+	pathWithoutZonePath := strings.Trim(snapshotIsiPath, zonePath)
+	directories := strings.Split(pathWithoutZonePath, "/")
 	// If there is no snapshot name in snapshot isi path or if it is empty
-	if len(dirs) < 4 || dirs[3] == "" {
+	if len(directories) < 2 || directories[2] == "" {
 		log.Debugf("invalid snapshot isilon path- '%s'", snapshotIsiPath)
 		return "", fmt.Errorf("invalid snapshot isilon path")
 	}
-
-	return dirs[3], nil
+	snapShotName = directories[2]
+	return snapShotName, nil
 }
 
-func (svc *isiService) GetSnapshotIsiPathComponents(snapshotIsiPath string) (string, string, string) {
+func (svc *isiService) GetSnapshotIsiPathComponents(snapshotIsiPath, zonePath string) (string, string, string) {
 	// Returns snapshot isi path components- isiPath, snapshotName, srcVolName
+	var isiPath string
+	var snapshotName string
+	// Snapshot isi path format /<ifs>/.snapshot/<snapshot_name>/<volume_path_without_ifs_prefix>
+	//Non System Access Zone snapshot isi path /<ifs>/<csi_zone_base_path>/.snapshot/<snapshot_name>/<volume_path_without_ifs_prefix>
 	dirs := strings.Split(snapshotIsiPath, "/")
-	isiPath := path.Join("/", dirs[1], strings.Join(dirs[4:len(dirs)-1], "/"))
-	snapshotName := dirs[3]
 	srcVolName := dirs[len(dirs)-1]
-
+	//in case of non system access zone
+	if dirs[2] != ".snapshot" {
```

Comment:
is this valid for isi path with more than 2 dir levels in it for ex: /ifs/dir1/dir2/dir3/ ?

Reply:
yes further nested if checks multiple directory path. With this if condition we are just checking if non system access zone .